### PR TITLE
HOTFIX: #26607 Fix failing CI due to Functional Tests

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
During fix of magento/magento2#26607 with magento/magento2#26608 - I've noticed that the test is extended and actually - the Admin session is killed, completely ignoring the fact, that further actions that are performed in original tests still need the session.

* Reorganized tests and it's extensions to work properly.
* Removed the `cache:flush` bad practice loved by Core Team <3
* Fixed issues caused by missing indexing (adding `cron:run --group=index`)
* Invalid search returns returned (duplicated products)
* Extended Credit Card expiration period to another 4 years

All above mentioned also for B2B and EE version.

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/135
https://github.com/magento/partners-magento2b2b/pull/45

### Fixed Issues (if relevant)
1. https://github.com/magento/magento2/issues/26607

### Manual testing scenarios (*)
1. Change README.md for any module
2. Submit PR to Github

### Questions or comments
:man_facepalming: 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml
Comment (ihor-sviziev) [inline app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: Why we need to add wait for 1 minute in a lot of tests? Tests run will just became longer, can’t we prevent it?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml
Comment (lbajsarowicz) [reply app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: It's temporary solution that will disappear when we will have <magentoCron command available (Monday). 

It is because of running different tests that run cron jobs at the same time. 

Can draw a graph explaining the issue.

Comment (ihor-sviziev) [reply app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: So I believe this is issue of mftf, Is there any link to it?

Comment (lbajsarowicz) [reply app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: Assume that when you run 4 workers of tests, and these tests are ran simultaneously - all 4 are running <createData and cron:run --group=index in 30 seconds. As a result - the first one gets the data indexer, but all the rest - not, as cron is waiting for 60 seconds before doing another round of indexing. That's the reason why it works randomly - it's just related with overlapping actions.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml
Comment (lbajsarowicz) [reply app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: I removed once again waiting for Indexers, maybe - by coincidence - this will work somehow.

Comment (lbajsarowicz) [reply app/code/Magento/Bundle/Test/Mftf/Test/StorefrontCustomerSelectAndSetBundleOptionsTest.xml]: @ihor-sviziev - The new feature for running Cron Jobs is mentioned in MFTF Changelog:
https://github.com/magento/magento2-functional-testing-framework/blob/master/CHANGELOG.md

https://github.com/magento/magento2-functional-testing-framework/pull/538 -- FEATURE: command to execute Cron Jobs

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

1 fixed = 1 new to be fixed :man_facepalming: 
And all the issues are similar and are result of invalid test isolation

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests CE

Reason:
image

**Comment by Nazar65** [comment]
> @magento run Functional Tests CE
> 
> Reason:
> image

Yes i had similar issue when i was trying to fix failed b2b on my PR

**Comment by lbajsarowicz** [comment]
Working on it - the structure of the tests related to searching products are broken.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

I've removed sample products that are not used for testing, as it breaks all Elasticsuite tests :man_facepalming:

**Comment by lbajsarowicz** [comment]
The issue of Tests Isolation is the issue, still.
The easiest way to avoid issues with searching by short description is to remove spaces from ShortDescription as it makes search catch other products.

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B

**Comment by lbajsarowicz** [comment]
I've rolled back to the best result so far and giving up.
This PR is passing the most of the issues that Magento CI is facing now, but I've spent 3 days to fix all of them and it's neverending story.

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests CE

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
Can't reproduce this issue:
image

Locally:
image

**Comment by lbajsarowicz** [comment]
Same with the issue:
image

Locally:
image

**Comment by lbajsarowicz** [comment]
> Can't reproduce this issue:
> image
> 
> Locally:
> image

OK. this is caused by the fact that Test Author seen no difference between `increment_id` and `entity_id`. Fixing that.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Replaced usage of `increment_id` :cry: to `entity_id`.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests
Maybe today!

**Comment by slavvka** [comment]
@lbajsarowicz almost :)

**Comment by lbajsarowicz** [comment]
@slavvka I've been told how Magento CI works currently. 
I should have fixed all tests to get it green, as the Suites are generated randomly :-(

**Review by ihor-sviziev — APPROVED** [review]
Tests have passed 🎉

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6825](https://jira.corp.magento.com/browse/ENGCOM-6825) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by lbajsarowicz** [comment]
@lenaorobei :+1: we can go with EE + B2B too.

**Comment by engcom-Delta** [comment]
**Notice:** QA not applicable

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@lbajsarowicz There are many MTF failures when running with MSI. Unfortunately currently the public builds (unlike internal builds) do not run tests with MSI. If you wish to address the failures you could try run all tests locally with MSI installed. I can show you the samples of the failures here:
```
no such element: Unable to locate element: {"method":"xpath","selector":"//div[@class='panel wrapper']//li/a[contains(.,'Create an Account')]"}
  (Session info: chrome=65.0.3325.162)
  (Driver info: chromedriver=2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752),platform=Linux 3.10.0-693.17.1.el7.x86_64 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '3.8.1', revision: '6e95a6684b', time: '2017-12-01T19:05:32.194Z'
System info: host: 'extfat.devops.magento.com', ip: '172.17.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '3.10.0-693.17.1.el7.x86_64', java.version: '1.8.0_181'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 2.36.540471 (9c759b81a907e7..., userDataDir: /tmp/.org.chromium.Chromium...}, cssSelectorsEnabled: true, databaseEnabled: false, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, rotatable: false, setWindowRect: true, takesHeapSnapshot: true, takesScreenshot: true, unexpectedAlertBehaviour: , unhandledPromptBehavior: , version: 65.0.3325.162, webStorageEnabled: true}
Session ID: 4a89a0851e8588ed0d061e0b1aa7d18a
*** Element info: {Using=xpath, value=//div[@class='panel wrapper']//li/a[contains(.,'Create an Account')]}
```
These failures are in:
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.AdminUserCreateOrderWithSimpleProductOnTestStockFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.LoggedInCustomerOrderedSimpleProductFromTestSourceUsingMultiShipmentsToSeveralAddressesTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontLoggedInCustomerCreateOrderWithSimpleProductOnTestStockFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.AdminUserCreateOrderWithSimpleProductWithManageStockNoInAdvancedInventorySettingsFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.AdminUserCreatedOrderWithVirtualProductOnTestStockFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.AdminUserCreatedOrderWithVirtualProductOnDefaultStockFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontLoggedInCustomerOrderedVirtualProductOnTestStockFromCustomWebsiteTest (CE, EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_EE_Multi_Mode.CreateOrderWithConfigurableProductCustomStockCustomWebsiteBySkuFromCustomerAccountTest (EE and B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_B2B_Multi_Mode.OrderWithMultipleConfigurableProductsAdditionalStock (B2B)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.OrderWithMultipleConfigurableProductsAdditionalStock (B2B)

Also there are other failures (maybe flaky but not sure):
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.ElasticSearchLayeredNavigationCustomStockTest  (CE)
```
Element located either by name, CSS or XPath element with '//a[contains(@href, 'price=10-20')]' was not found.
```
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.ReorderButtonGroupedProductOnTestStockTest  (CE)
```
invalid element state: Element is not currently interactable and may not be manipulated
  (Session info: chrome=65.0.3325.162)
  (Driver info: chromedriver=2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752),platform=Linux 3.10.0-693.17.1.el7.x86_64 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds
Build info: version: '3.8.1', revision: '6e95a6684b', time: '2017-12-01T19:05:32.194Z'
System info: host: 'extfat.devops.magento.com', ip: '172.17.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '3.10.0-693.17.1.el7.x86_64', java.version: '1.8.0_181'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 2.36.540471 (9c759b81a907e7..., userDataDir: /tmp/.org.chromium.Chromium...}, cssSelectorsEnabled: true, databaseEnabled: false, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, rotatable: false, setWindowRect: true, takesHeapSnapshot: true, takesScreenshot: true, unexpectedAlertBehaviour: , unhandledPromptBehavior: , version: 65.0.3325.162, webStorageEnabled: true}
Session ID: f33912354205a275906b6fbc6e4e6aec
```
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontButtonReorderForTheOrderWithConfigurableProductOnTestStockInCustomerCabinetOnHomepageIsNotPresentIfProductSoldOutTest  (CE)
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontButtonReorderForTheOrderWithConfigurableProductOnDefaultStockInCustomerCabinetOnHomepageIsNotPresentIfProductSoldOutTest  (CE)
-   Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontButtonReorderForTheOrderWithConfigurableProductOnDefaultStockInCustomerCabinetOnHomepageIsNotPresentIfProductSoldOutTest  (CE)
```
Link or Button or CSS or XPath element with 'button.action-default.save.submit-button' was not found.
```
-  Magento\FunctionalTestingFramework.functional\MSI_Multi_Mode.StorefrontQuickOrderSimpleProductsCustomStockTest (B2B)
```
invalid element state: Element is not currently interactable and may not be manipulated
  (Session info: chrome=65.0.3325.162)
  (Driver info: chromedriver=2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752),platform=Linux 3.10.0-693.17.1.el7.x86_64 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds
Build info: version: '3.8.1', revision: '6e95a6684b', time: '2017-12-01T19:05:32.194Z'
System info: host: 'extfat.devops.magento.com', ip: '172.17.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '3.10.0-693.17.1.el7.x86_64', java.version: '1.8.0_181'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 2.36.540471 (9c759b81a907e7..., userDataDir: /tmp/.org.chromium.Chromium...}, cssSelectorsEnabled: true, databaseEnabled: false, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, rotatable: false, setWindowRect: true, takesHeapSnapshot: true, takesScreenshot: true, unexpectedAlertBehaviour: , unhandledPromptBehavior: , version: 65.0.3325.162, webStorageEnabled: true}
Session ID: 24201a0c20f97933745380d415e25755
```

We are working on adding MSI to the public builds as it is a required part of the internal builds. Also we will add a possibility to link MSI RPs as related. Sorry for inconvenience!

**Comment by lbajsarowicz** [comment]
@slavvka I haven't changed these selectors in my PR. 
Why should I cover that as a part of this PR?

**Comment by slavvka** [comment]
@lbajsarowicz I didn't say you should. I said if you wish. Otherwise the code cannot be merged to the mainline as it fails the builds. I do know that it is our issue that MSI tests aren't run on the public builds but that is what it is. I already added `magento/inventory` as related to `magento/magento2` so I think we can create a related MSI PR here and run the tests to see and fix the issues. If you wish of course

**Comment by slavvka** [comment]
I linked this PR to a test one in MSI

**Comment by slavvka** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
I can't even install MSI :-/
`magento/inventory-composer-metapackage-dev`
It's not available.
```
php .ee/dev/tools/build-ee.php --ce-source . --ee-source .inventory 
```

also fails (missing modules).

**Comment by lbajsarowicz** [comment]
Let's merge working atomic parts, and then wonder about further steps.
Currently, we spend 90% of the time to wait and repeat tests... instead of stabilizing them one by one

**Comment by slavvka** [comment]
@lbajsarowicz As you wish. BTW you can see the failing tests now.
Regarding flaky tests I can say that they got much stable after all I merged during the last week.

**Comment by lbajsarowicz** [comment]
> @lbajsarowicz As you wish. BTW you can see the failing tests now.
> Regarding flaky tests I can say that they got much stable after all I merged during the last week.

I tried to run these tests with EE + B2B + PageBuilder + PageBuilder EE... and it does not even generate tests properly. It's too much work for 1 person to fix whole Magento :-( 

I'm doing that from May 2019 when it comes to Functional Tests, and it's really huge effort put to stabilize, understand, run that.

**Comment by slavvka** [comment]
Not PB. MSI. But I completely understand that. I will ask our guys to help you. This PR is very critical

**Comment by slavvka** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@slavvka I see the reason! Will try to fix.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@lbajsarowicz The failures look like flaky but I cannot see them in the delivery so if you could try to fix them I would appreciate :)

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests
(After changes to Inventory and Magento branches)

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
There were conflicts in `ee` PR, now resolved.

**Comment by lenaorobei** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by lenaorobei** [comment]
@lbajsarowicz failures are not related to your changes. 
But since MFTF `2.6.1` is already in mainline, could you please change your fixes to use `magentoCron`?

**Comment by lbajsarowicz** [comment]
@lenaorobei  I'd like to do that for whole Magento in another PR if you don't mind. That would be easier for me to maintain. I'm just fed up with PRs hanging for 3 months (like the one with extracting Customer Action Groups). It makes lots of mess, conflicts etc.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6825](https://jira.corp.magento.com/browse/ENGCOM-6825) has been created to process this Pull Request

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests
:pray: :pray: :pray:

**Comment by zakdma** [comment]
@lbajsarowicz  it is bad practice to hide issues adding reindexing everywhere even it is not necessary. You claim core team but do actually the same.

**Comment by lbajsarowicz** [comment]
@zakdma 
1. I am not running reindexes
2. I am running Cron Jobs - which is exactly the same thing that happens on Production Environments

Cron Jobs from `index` group are responsible for updating single entities indexes based on `_cl` tables that are created once you configure reindex to run `on schedule` instead of `realtime`.

I'd love to know your point of view.

**Comment by zakdma** [comment]
@lbajsarowicz  it is bad practice to hide issues adding reindexing everywhere even it is not necessary. You claim core team but do actually the same.
Also there is special AG for this purposes
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Indexer/Test/Mftf/ActionGroup/CliRunReindexUsingCronJobsActionGroup.xml

**Comment by lbajsarowicz** [comment]
I don't understand @zakdma why - in your opinion - running Cron Jobs is a bad practice?

The Action Group you provided is not working as expected, because if you run `cron:run` twice, and it will be executed in the same minute - second one is going to be skipped.

**Comment by zakdma** [comment]
No, you are not right. First run schedule jobs and second one run scheduled jobs.

**Comment by zakdma** [comment]
cron:run --group=index actually just do reindex of broken indices. If indices are not broken this is not necessary to run it. Indices are broken after, for instance, creating product attribute. That is expected behaviour and we need to reindex them after that. If previous test that create attribute ran before test that depends on indices it will fail but issue not failed test but test before it.

**Comment by lbajsarowicz** [comment]
There is a simple way to prove your ActionGroup does not work as expected:

Run:
```bash
watch -n5 "bin/magento cron:run --group=index -vvv"
```

And then watch the logs after several minutes:
```
[2020-02-24 12:02:38] report.INFO: Cron Job indexer_reindex_all_invalid is run [] []
[2020-02-24 12:02:39] report.INFO: Cron Job indexer_reindex_all_invalid is successfully finished. Statistics: {"sum":0.32009887695312,"count":1,"realmem":14680064,"emalloc":16307456,"realmem_start":48758784,"emalloc_start":45042280} [] []
[2020-02-24 12:02:39] report.INFO: Cron Job indexer_update_all_views is run [] []
[2020-02-24 12:02:39] report.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":0.40363311767578,"count":1,"realmem":10485760,"emalloc":6242624,"realmem_start":63438848,"emalloc_start":61352328} [] []
[2020-02-24 12:03:06] report.INFO: Cron Job indexer_reindex_all_invalid is run [] []
[2020-02-24 12:03:06] report.INFO: Cron Job indexer_reindex_all_invalid is successfully finished. Statistics: {"sum":0.015038967132568,"count":1,"realmem":0,"emalloc":905128,"realmem_start":46661632,"emalloc_start":44110200} [] []
[2020-02-24 12:03:06] report.INFO: Cron Job indexer_update_all_views is run [] []
[2020-02-24 12:03:06] report.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":0.1584370136261,"count":1,"realmem":6291456,"emalloc":6716520,"realmem_start":46661632,"emalloc_start":45017920} [] []
[2020-02-24 12:04:00] report.INFO: Cron Job indexer_reindex_all_invalid is run [] []
[2020-02-24 12:04:00] report.INFO: Cron Job indexer_reindex_all_invalid is successfully finished. Statistics: {"sum":0.014179229736328,"count":1,"realmem":0,"emalloc":905128,"realmem_start":46661632,"emalloc_start":44110200} [] []
[2020-02-24 12:04:00] report.INFO: Cron Job indexer_update_all_views is run [] []
[2020-02-24 12:04:00] report.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":0.17619705200195,"count":1,"realmem":6291456,"emalloc":6716520,"realmem_start":46661632,"emalloc_start":45017920} [] []
[2020-02-24 12:05:05] report.INFO: Cron Job indexer_reindex_all_invalid is run [] []
[2020-02-24 12:05:05] report.INFO: Cron Job indexer_reindex_all_invalid is successfully finished. Statistics: {"sum":0.012212991714478,"count":1,"realmem":0,"emalloc":905128,"realmem_start":46661632,"emalloc_start":44110200} [] []
[2020-02-24 12:05:05] report.INFO: Cron Job indexer_update_all_views is run [] []
[2020-02-24 12:05:05] report.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":0.14868688583374,"count":1,"realmem":6291456,"emalloc":6716520,"realmem_start":46661632,"emalloc_start":45017920} [] []
[2020-02-24 12:06:04] report.INFO: Cron Job indexer_reindex_all_invalid is run [] []
[2020-02-24 12:06:04] report.INFO: Cron Job indexer_reindex_all_invalid is successfully finished. Statistics: {"sum":0.016084909439087,"count":1,"realmem":0,"emalloc":905128,"realmem_start":46661632,"emalloc_start":44110200} [] []
[2020-02-24 12:06:04] report.INFO: Cron Job indexer_update_all_views is run [] []
[2020-02-24 12:06:04] report.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":0.17505884170532,"count":1,"realmem":6291456,"emalloc":6716520,"realmem_start":46661632,"emalloc_start":45017920} [] []
```

I'd love to learn something new - so let me understand your point of view.

**Comment by lbajsarowicz** [comment]
@zakdma This is actually sad, that you are throwing stones in my direction without even checking the triggers that are declared in `mview.xml`.

As an example 
https://github.com/magento/magento2/blob/9fe3b8a0cdff5c9f4677c1890ca2bf52718acae5/app/code/Magento/Catalog/etc/mview.xml#L54-L60

Creates trigger on MySQL:
```sql
create or replace definer = magento@`%` trigger trg_catalog_product_entity_decimal_after_delete
	after delete
	on catalog_product_entity_decimal
	for each row
	BEGIN
SET @entity_id = (SELECT `entity_id` FROM `catalog_product_entity` WHERE `row_id` = OLD.`row_id`);
INSERT IGNORE INTO `catalogsearch_fulltext_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalog_product_attribute_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalogrule_product_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalog_product_price_cl` (`entity_id`) values(@entity_id);
END;

create or replace definer = magento@`%` trigger trg_catalog_product_entity_decimal_after_insert
	after insert
	on catalog_product_entity_decimal
	for each row
	BEGIN
SET @entity_id = (SELECT `entity_id` FROM `catalog_product_entity` WHERE `row_id` = NEW.`row_id`);
INSERT IGNORE INTO `catalogsearch_fulltext_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalog_product_attribute_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalogrule_product_cl` (`entity_id`) values(@entity_id);
INSERT IGNORE INTO `catalog_product_price_cl` (`entity_id`) values(@entity_id);
END;

create or replace definer = magento@`%` trigger trg_catalog_product_entity_decimal_after_update
	after update
	on catalog_product_entity_decimal
	for each row
	BEGIN
SET @entity_id = (SELECT `entity_id` FROM `catalog_product_entity` WHERE `row_id` = NEW.`row_id`);
IF (NOT(NEW.`value_id` <=> OLD.`value_id`) OR NOT(NEW.`attribute_id` <=> OLD.`attribute_id`) OR NOT(NEW.`store_id` <=> OLD.`store_id`) OR NOT(NEW.`value` <=> OLD.`value`) OR NOT(NEW.`row_id` <=> OLD.`row_id`)) THEN INSERT IGNORE INTO `catalogsearch_fulltext_cl` (`entity_id`) values(@entity_id); END IF;
IF (NOT(NEW.`value_id` <=> OLD.`value_id`) OR NOT(NEW.`attribute_id` <=> OLD.`attribute_id`) OR NOT(NEW.`store_id` <=> OLD.`store_id`) OR NOT(NEW.`value` <=> OLD.`value`) OR NOT(NEW.`row_id` <=> OLD.`row_id`)) THEN INSERT IGNORE INTO `catalog_product_attribute_cl` (`entity_id`) values(@entity_id); END IF;
IF (NOT(NEW.`value_id` <=> OLD.`value_id`) OR NOT(NEW.`attribute_id` <=> OLD.`attribute_id`) OR NOT(NEW.`store_id` <=> OLD.`store_id`) OR NOT(NEW.`value` <=> OLD.`value`) OR NOT(NEW.`row_id` <=> OLD.`row_id`)) THEN INSERT IGNORE INTO `catalogrule_product_cl` (`entity_id`) values(@entity_id); END IF;
IF (NOT(NEW.`value_id` <=> OLD.`value_id`) OR NOT(NEW.`attribute_id` <=> OLD.`attribute_id`) OR NOT(NEW.`store_id` <=> OLD.`store_id`) OR NOT(NEW.`value` <=> OLD.`value`) OR NOT(NEW.`row_id` <=> OLD.`row_id`)) THEN INSERT IGNORE INTO `catalog_product_price_cl` (`entity_id`) values(@entity_id); END IF;
END;
```

As a result, the `entity_id` is being saved to `catalogsearch_fulltext_cl` (and 3 others).

The indexer is looking for changes in these tables to update indexes for single entities.

**Comment by zakdma** [comment]
I think we don't understand each other.
1) MFTF Tests run on environment with cron disabled
2) First manual cron:run (on an clear database) will not run jobs but just schedule jobs
Second manual call cron:run will actually run jobs even if it run in the same minute with first call
3) --group=index will run job indexer_reindex_all_invalid that not just index changes (that created in tables) but totally reindex all broken indices (invalidated indices)

**Comment by lbajsarowicz** [comment]
1. That is why we need to run them manually
2. I've never had such issue. Previous `<magentoCLI command="cron:run --group=index"...>` and current `<magentoCron groups="index">` works as expected. The only exception to `<magentoCLI` was the described case of Crons running on the same minute, especially for Functional tests running simultaneously. 
3. It is going to run 3 jobs:
```xml
<group id="index">
        <job name="indexer_reindex_all_invalid" instance="Magento\Indexer\Cron\ReindexAllInvalid" method="execute">
            <schedule>* * * * *</schedule>
        </job>
        <job name="indexer_update_all_views" instance="Magento\Indexer\Cron\UpdateMview" method="execute">
            <schedule>* * * * *</schedule>
        </job>
        <job name="indexer_clean_all_changelogs" instance="Magento\Indexer\Cron\ClearChangelog" method="execute">
            <schedule>0 * * * *</schedule>
        </job>
    </group>
```

* `indexer_reindex_all_invalid` is going to run for `invalid` indices
* `indexer_update_all_views`  is going to run for outdated indices

Still, I have no idea why are you throwing stones in my direction saying that I hide the issues. As you see above: It's necessary to run Cron for `indexer_update_all_views` every single time when entities are modified **and** indexers are configured to work on schedule.

**Comment by zakdma** [comment]
2) It is not issue but normal behaviour
3) But for MFTF tests make sense only indexer_reindex_all_invalid because MFTF tests env don't use triggers and tables with changes since it works in Update on Save mode

> It's necessary to run Cron for indexer_update_all_views every single time when entities are modified and indexers are configured to work on schedule.
Again, MFTF test env works in  Update on Save mode for now.

About hiding. It better to know what actually make index invalid. Maybe it is a bug that will be hidden if  you reindex invalidated indices every time you change something. For Ex. we found bug with invalidated indices during fixing such "flaky" MFTF test. We could just add reindex and go ahead but we make deep investigation and finally found it. Now we have tons of tests "fixed" with reindex and hidden real cause of unexpected behaviour. That is sad.

**Comment by lbajsarowicz** [comment]
3. Actually, whether MFTF should run `On schedule` or `On save` should be defined at the very beginning of each suite. Let's consider that both way:
   a) Reindex `On save` (`realtime`): `cron:run --group=index` does nothing
   b) Reindex `On schedule`: `cron:run --group=index` runs `indexer_update_all_views`

That is why after `createData`, `updateData` and to be nice - even after `deleteData` you should run index crons.

**Comment by zakdma** [comment]
> That is why we have Integration Tests, actually.

Not all code covered by IT.
It is always better to have more than one approach to test.

**Comment by lbajsarowicz** [comment]
There are no obstacles for you to contribute with your approach.
Ha! You are more than welcome to stabilize the Functional Tests.

**Comment by zakdma** [comment]
Actually I do it almost every single day.

**Pull request merged**
